library/init/meta/interactive.lean

```diff
@@ -521,6 +517,25 @@ do x ← mk_fresh_name,
    t ← get_local x,
    induction (to_pexpr t) rec_name hs ([] : list name)
 
+private meta def cases_arg_p : parser (option name × pexpr) :=
+with_desc "(id :)? expr" $ do
+  t ← texpr,
+  match t with
+  | (local_const x _ _ _) :=
```

Comment:
The term-mode `have` will default to interpreting it as a type, so it should be reasonable to do the same in tactic mode.